connectors/sources/graphql.py

```diff
@@ -0,0 +1,519 @@
+                "type": "str",
+                "required": False,
+            },
+            "graphql_object_list": {
+                "label": "GraphQL Objects List",
+                "order": 9,
+                "tooltip": "Specifies which GraphQL objects should be indexed as individual documents. This allows finer control over indexing, ensuring only relevant data sections from the GraphQL response are stored as separate documents. Use '.' to provide full path of the object from the root of the response. For example 'organization.users.nodes'",
+                "type": "list",
+            },
+            "headers": {
+                "label": "Headers",
+                "order": 10,
+                "type": "str",
+                "required": False,
+            },
+            "pagination_model": {
+                "display": "dropdown",
+                "label": "Pagination model",
+                "options": [
+                    {"label": "No pagination", "value": NO_PAGINATION},
+                    {"label": "Cursor-based pagination", "value": CURSOR_PAGINATION},
+                ],
+                "order": 11,
+                "tooltip": "Cursor based pagination requires 'pageInfo' field along with argument 'after' for objects mentioned in 'GraphQL Objects List'. It also requires variable for 'after' argument.",
+                "type": "str",
+                "value": NO_PAGINATION,
+            },
+            "pagination_key": {
+                "depends_on": [
+                    {"field": "pagination_model", "value": CURSOR_PAGINATION}
+                ],
+                "label": "Pagination key",
+                "order": 12,
+                "tooltip": "Specifies which GraphQL object is used for pagination. Use '.' to provide full path of the object from the root of the response. For example 'organization.users'",
+                "type": "str",
+            },
+            "connection_timeout": {
+                "default_value": 300,
+                "display": "numeric",
+                "label": "Connection Timeout",
+                "order": 13,
+                "required": False,
+                "type": "int",
+                "ui_restrictions": ["advanced"],
+            },
+        }
+
+    def is_query(self, graphql_query):
+        try:
+            ast = parse(graphql_query)  # pyright: ignore
+            for definition in ast.definitions:  # pyright: ignore
+                if (
+                    hasattr(definition, "operation")
+                    and definition.operation.value != "query"
+                ):
+                    return False
+            return True
+        except Exception as e:
+            self._logger.error(f"Failed to parse GraphQL query. Error: {e}")
+            return False
+
+    def validate_endpoints(self):
+        if re.match(URL_REGEX, self.graphql_client.url):
+            return True
+        return False
+
+    async def validate_config(self):
```

Comment:
Small ask: since we can build query ast at this stage already, can we validate that:
1. graphql objects list (. separated) are defined in the graphql query?
2. if pagination is enabled: pagination key (. separated) is a valid path and contains a pageInfo reference?

Does it make sense? My idea is to fail fast and give the feedback quickly in case this is miscomfigured

Reply:
I tried to think in that direction and attempted to implement it. However, it is not possible because in AST, all the nodes are considered fields. But for a dot-separated object, we need to check if the field is a dictionary as well, which we are doing here.(https://github.com/elastic/connectors/pull/2134/files#diff-1ac50fbc6d32ff7dfa228d41ad1b8945bd951532d30582faf148690d642ba397R159)